File: LeetCode/33.search-in-rotated-sorted-array.cpp
/*
 * @lc app=leetcode id=33 lang=cpp
 *
 * [33] Search in Rotated Sorted Array
 *
 * https://leetcode.com/problems/search-in-rotated-sorted-array/description/
 *
 * algorithms
 * Medium (32.53%)
 * Total Accepted:    354.2K
 * Total Submissions: 1.1M
 * Testcase Example:  '[4,5,6,7,0,1,2]\n0'
 *
 * Suppose an array sorted in ascending order is rotated at some pivot unknown
 * to you beforehand.
 * 
 * (i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).
 * 
 * You are given a target value to search. If found in the array return its
 * index, otherwise return -1.
 * 
 * You may assume no duplicate exists in the array.
 * 
 * Your algorithm's runtime complexity must be in the order of O(log n).
 * 
 * Example 1:
 * 
 * 
 * Input: nums = [4,5,6,7,0,1,2], target = 0
 * Output: 4
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: nums = [4,5,6,7,0,1,2], target = 3
 * Output: -1
 * 
 */
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int len = nums.size();
        int left = 0, right = len-1;
        int left_end = -1, right_begin = 0;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (mid-1 >= 0 && nums[mid-1] > nums[mid]) {
                left_end = mid-1;
                right_begin = mid;
                break;
            }
            if (mid+1 < len && nums[mid] > nums[mid+1]) {
                left_end = mid;
                right_begin = mid+1;
                break;
            }
            if (nums[mid] > nums[0]) {
                left = mid+1;
            } else {
                right = mid-1;
            }
        }
        int res1 = binarySearch(nums, 0, left_end, target);
        int res2 = binarySearch(nums, right_begin, nums.size()-1, target);

        if (res1 != -1) return res1;
        if (res2 != -1) return res2;
        return -1;
    }

private:
    int binarySearch(vector<int>& nums, int left, int right, int target) {
        while (left <= right) {
            int mid = left + (right - left) / 2;
            int midNum = nums[mid];
            if (target < midNum) right = mid-1;
            else if (target > midNum) left = mid+1;
            else return mid;
        }

        return -1;
    }
};

File: java/33.search-in-rotated-sorted-array.java
/*
 * @lc app=leetcode id=33 lang=java
 *
 * [33] Search in Rotated Sorted Array
 *
 * https://leetcode.com/problems/search-in-rotated-sorted-array/description/
 *
 * algorithms
 * Medium (37.57%)
 * Likes:    13180
 * Dislikes: 863
 * Total Accepted:    1.4M
 * Total Submissions: 3.7M
 * Testcase Example:  '[4,5,6,7,0,1,2]\n0'
 *
 * There is an integer array nums sorted in ascending order (with distinct
 * values).
 * 
 * Prior to being passed to your function, nums is possibly rotated at an
 * unknown pivot index k (1 <= k < nums.length) such that the resulting array
 * is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]]
 * (0-indexed). For example, [0,1,2,4,5,6,7] might be rotated at pivot index 3
 * and become [4,5,6,7,0,1,2].
 * 
 * Given the array nums after the possible rotation and an integer target,
 * return the index of target if it is in nums, or -1 if it is not in nums.
 * 
 * You must write an algorithm with O(log n) runtime complexity.
 * 
 * 
 * Example 1:
 * Input: nums = [4,5,6,7,0,1,2], target = 0
 * Output: 4
 * Example 2:
 * Input: nums = [4,5,6,7,0,1,2], target = 3
 * Output: -1
 * Example 3:
 * Input: nums = [1], target = 0
 * Output: -1
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= nums.length <= 5000
 * -10^4 <= nums[i] <= 10^4
 * All values of nums are unique.
 * nums is an ascending array that is possibly rotated.
 * -10^4 <= target <= 10^4
 * 
 * 
 */

// @lc code=start
class Solution {
    public int search(int[] nums, int target) {
        int len = nums.length;
        int minIdx = findMinIdx(nums);
        int res = binarySearch(nums, 0, minIdx-1, target);
        if (res == -1) {
            res = binarySearch(nums, minIdx, len-1, target);
        }
        return res;
    }

    public int findMinIdx(int[] nums) {
        int len = nums.length, s = 0, e = len - 1;
        while (s < e) {
            int mid = (s+e)/2;
            int m = nums[mid];

            if (m >= nums[s] && m > nums[e]) {
                s = mid + 1;
            } else {
                e = mid;
            }
        }
        return s;
    }

    public int binarySearch(int[] nums, int s, int e, int target) {
        if (s > e) {
            return -1;
        }

        while (s <= e) {
            int mid = (s+e)/2;
            int m = nums[mid];

            if (m < target) {
                s = mid + 1;
            } else if (m == target) {
                return mid;
            } else {
                e = mid - 1;
            }
        }

        return -1;
    }
}
// @lc code=end

